dids/did_dht.go

```diff
@@ -0,0 +1,314 @@
+package dids
+
+import (
+	"encoding/base64"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"golang.org/x/net/dns/dnsmessage"
+
+	"github.com/tbd54566975/web5-go/crypto/dsa"
+	"github.com/tbd54566975/web5-go/jwk"
+	"github.com/tv42/zbase32"
+)
+
+var txtEntityNames = map[string]struct{}{
+	"vm":   {},
+	"auth": {},
+	"asm":  {},
+	"agm":  {},
+	"inv":  {},
+	"del":  {},
+}
+
+var relationshipDNStoDID = map[string]string{
+	"auth": "authentication",
+	"asm":  "assertionMethod",
+	"agm":  "keyAgreement",
+	"inv":  "capabilityInvocation",
+	"del":  "capabilityDelegation",
+}
+
+var keyTypes = map[string]string{
+	"0": dsa.AlgorithmIDED25519,
+	"1": dsa.AlgorithmIDSECP256K1,
+	"2": dsa.AlgorithmIDSECP256K1,
+}
+
+// DHT Resolver is used to create DHT DNS representations of DID documents and to resolve
+// DNS record serialized DID documents
+type DHTResolver struct {
+	client *http.Client
+}
+
+// dhtDidRecord is used to structure the DNS representation of a DID
+type dhtDidRecord struct {
+	rootRecord string
+	records    map[string]string
+}
+
+func (rec *dhtDidRecord) DIDDocument() *Document {
+	relationshipMap := parseVerificationRelationships(rec.rootRecord)
+
+	// Now we have a did in a dns record. yay
+	document := &Document{}
+
+	// Now create the did document
+	for name, data := range rec.records {
+
+		switch {
+		case strings.HasPrefix(name, "_k"):
+			{
+				vMethod, err := UnmarshalVerificaitonMethod(data)
+				if err != nil {
+					// TODO handle error
+				}
+				entryId := strings.Split(name, ".")[0]
+				relationships, ok := relationshipMap[entryId]
+
+				if !ok {
+					// no relationships
+					continue
+				}
+
+				opts := []string{}
+				for _, r := range relationships {
+					if o, ok := relationshipDNStoDID[r]; ok {
+						opts = append(opts, o)
+					}
+				}
+
+				document.AddVerificationMethod(
+					*vMethod,
+					Purposes(opts...),
+				)
+			}
+		case strings.HasPrefix(name, "_s"):
+			s := UnmarshalService(data)
+			document.AddService(s)
+		case strings.HasPrefix(name, "_cnt"):
+			// TODO add controller https://did-dht.com/#controller
+			// optional field
+			// comma-separated list of controller DID identifiers
+			document.Controller = strings.Split(data, ",")
+		case strings.HasPrefix(name, "_aka"):
+			// TODO add aka https://did-dht.com/#also-known-as
+			document.AlsoKnownAs = strings.Split(data, ",")
+		default:
+		}
+	}
+
+	return document
+}
+
+func (r *DHTResolver) Resolve(uri, relay string) (*Document, error) {
```

Comment:
* i think we'll want this to be named `ResolveDIDDHT` to match the pattern set for `did:jwk`?
* do we need `DHTResolver`? can we allow an http client to be passed in optionally and create one by default otherwise? we'll need to think thru how to "register" did resolvers into the default resolver [here](https://github.com/TBD54566975/web5-go/blob/main/dids/resolver.go)

Reply:
(1) yep. will rename. My initial intuition was that there might be opportunity for an interface if both methods have the same signature, but I have looked hard enough at the similarities to see if it's possible to extract them.
(2) to my point above, the JWK version doesn't need an http client; so this "pattern" would make the resolve method more similar to the jwk one; because the Resolver or the http client could be initialized with the proper relay. I can think about this and propose the updates in a subsequent commit.